[libc++][lit] Allow overriding the executor for tests

arichardson opened the pull request:
This is useful when trying to run multiple tests with different
arguments to the executor script. This is needed in my case since I
do not know the correct ssh connection arguments when building libc++.
The testing script I have spawns multiple QEMU instances that listen on
a given port on localhost and runs lit with the --num-shards/--run-shard
argument. In order to connect each shard to the right QEMU instances I
need to customize the arguments passed to ssh.py (--extra-ssh-args and
--extra-scp-args) but can't do this at configure time since the target
port is only known when running the tests but not when calling CMake.
This change allows me to pass `executor=ssh.py <args>` to lit once I know
the right hostname/port for running the tests.

This also deprecates the `LIB{CXX,CXXABI,UNWIND}_EXECUTOR` CMake variable
as the same can be achieved by adding `executor=...` to the
`LIB{CXX,CXXABI,UNWIND}_TEST_PARAMS` variable.

ldionne commented on libcxx/utils/libcxx/test/params.py:
```diff
@@ -330,5 +330,14 @@ def getModuleFlag(cfg, enable_modules):
+    Parameter(
+        name="executor",
+        type=str,
+        default="",
```
This default value could be `python3 <path-to>/run.py`.

ldionne commented on libcxx/utils/libcxx/test/params.py:
```diff
@@ -330,5 +330,14 @@ def getModuleFlag(cfg, enable_modules):
             AddCompileFlag("-D_LIBCPP_REMOVE_TRANSITIVE_INCLUDES"),
         ],
     ),
+    Parameter(
```
We could potentially remove `LIBCXX_EXECUTOR` if we have this setting? Then people could use `LIBCXX_TEST_PARAMS` to customize the executor.

It would be really nice to drop CMake options that are only related to testing, since we can pass them as Lit parameters.

If we do that, we'd want to add an entry to `releasenotes/18.rst` probably under Build System Changes.

arichardson replied:
That sounds like a great idea. I'll submit a follow-up change for this.

ldionne replied:
I think we should make it part of this change.

EricWF commented:
Looks fine to me. I would like @ldionne approval though.

ldionne reviewed (changes requested):
I think this makes a lot of sense! I have a few suggestions but I definitely like the direction.

arichardson committed:
[libunwind][libc++][libc++abi] Allow overriding the executor for tests

This is useful when trying to run multiple tests with different
arguments to the executor script. This is needed in my case since I
do not know the correct ssh connection arguments when building libc++.
The testing script I have spawns multiple QEMU instances that listen on
a given port on localhost and runs lit with the --num-shards/--run-shard
argument. In order to connect each shard to the right QEMU instances I
need to customize the arguments passed to ssh.py (--extra-ssh-args and
--extra-scp-args) but can't do this at configure time since the target
port is only known when running the tests but not when calling CMake.
This change allows me to pass `executor=ssh.py <args>` to lit once I know
the right hostname/port for running the tests.

This also deprecates the `LIB{CXX,CXXABI,UNWIND}_EXECUTOR` CMake variable
as the same can be achieved by adding `executor=...` to the
`LIB{CXX,CXXABI,UNWIND}_TEST_PARAMS` variable.

arichardson commented on libcxx/docs/ReleaseNotes/18.rst:
```diff
@@ -134,3 +134,8 @@ ABI Affecting Changes
 --------------------
+
+- The ``LIBCXX_EXECUTOR`` CMake variable has been deprecated. If you are relying on this, the new replacement
+  is either passing ``-Dexecutor=...`` to `llvm-lit`` or to make it persistent in the generated configuration
```
```suggestion
  is either passing ``-Dexecutor=...`` to ``llvm-lit`` or to make it persistent in the generated configuration
```

ldionne commented on libcxx/utils/libcxx/test/params.py:
```diff
@@ -314,5 +317,14 @@ def getStdFlag(cfg, std):
+        type=str,
+        default=f"{sys.executable} {Path(__file__).resolve().parent.parent.parent / 'run.py'}",
+        help="Custom executor to use instead of the configured default.",
+        actions=lambda executor: [] if not executor else [
```
There *has* to be an executor, so I think we should change to

```
actions=lambda executor: [AddSubstitution("%{executor}", executor)]
```

i.e. drop the `if`.

arichardson replied:
Thanks! Will merge once CI completes successfully.

ldionne reviewed (approved):
I really like this! Minor nitpick in `params.py` but this basically LGTM.

vvereschaka commented:
@arichardson 

here is a problem with your changes when building the libraries on the windows cross toolchain builders
* https://lab.llvm.org/buildbot/#/builders/60/builds/14086
* https://lab.llvm.org/buildbot/#/builders/119/builds/15283

https://lab.llvm.org/buildbot/#/builders/60/builds/14086/steps/12/logs/stdio
```
llvm-lit.py: C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\utils\lit\lit\TestingConfig.py:151: fatal: unable to parse config file 'C:/buildbot/as-builder-1/x-armv7l/build/runtimes/runtimes-armv7-unknown-linux-gnueabihf-bins/libunwind/test/cmake-bridge.cfg', traceback: Traceback (most recent call last):
  File "C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\utils\lit\lit\TestingConfig.py", line 139, in load_from_path
    exec(compile(data, path, "exec"), cfg_globals, None)
  File "C:/buildbot/as-builder-1/x-armv7l/build/runtimes/runtimes-armv7-unknown-linux-gnueabihf-bins/libunwind/test/cmake-bridge.cfg", line 7
    config. 'executor="C:/Python310/python.exe" "C:/buildbot/as-builder-1/x-armv7l/llvm-project/libcxx/utils/ssh.py" --host = "ubuntu@jetson6.lab.llvm.org'"
                                                                                                                                                           ^
SyntaxError: unterminated string literal (detected at line 7)

```
would you take care of it?

arichardson commented:
> @arichardson
> 
> here is a problem with your changes when building the libraries on the windows cross toolchain builders
> 
> * https://lab.llvm.org/buildbot/#/builders/60/builds/14086
> * https://lab.llvm.org/buildbot/#/builders/119/builds/15283
> 
> https://lab.llvm.org/buildbot/#/builders/60/builds/14086/steps/12/logs/stdio
> 
> ```
> llvm-lit.py: C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\utils\lit\lit\TestingConfig.py:151: fatal: unable to parse config file 'C:/buildbot/as-builder-1/x-armv7l/build/runtimes/runtimes-armv7-unknown-linux-gnueabihf-bins/libunwind/test/cmake-bridge.cfg', traceback: Traceback (most recent call last):
>   File "C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\utils\lit\lit\TestingConfig.py", line 139, in load_from_path
>     exec(compile(data, path, "exec"), cfg_globals, None)
>   File "C:/buildbot/as-builder-1/x-armv7l/build/runtimes/runtimes-armv7-unknown-linux-gnueabihf-bins/libunwind/test/cmake-bridge.cfg", line 7
>     config. 'executor="C:/Python310/python.exe" "C:/buildbot/as-builder-1/x-armv7l/llvm-project/libcxx/utils/ssh.py" --host = "ubuntu@jetson6.lab.llvm.org'"
>                                                                                                                                                            ^
> SyntaxError: unterminated string literal (detected at line 7)
> ```
> 
> would you take care of it?
> 
> UPDATE: Just let me know if you will need to check you changes before commit. I will test them on the builder locally.

Sorry about that! It looks like the cmake file quoting needs to be updated. Alternatively I could revert the part of this change that deprecates the executor variable. I will take a look at this first thing tomorrow morning. If it's more urgent, feel free to revert the change.

ldionne commented:
> > @arichardson
>
> Sorry about that! It looks like the cmake file quoting needs to be updated. Alternatively I could revert the part of this change that deprecates the executor variable.

Let's not un-deprecate the executor variables -- it's really great that we're removing some `LIBCXX_FOO` variables from the CMake build if they are only used for testing.

arichardson commented:
> @arichardson
> 
> here is a problem with your changes when building the libraries on the windows cross toolchain builders
> 
> * https://lab.llvm.org/buildbot/#/builders/60/builds/14086
> * https://lab.llvm.org/buildbot/#/builders/119/builds/15283
> 
> https://lab.llvm.org/buildbot/#/builders/60/builds/14086/steps/12/logs/stdio
> 
> ```
> llvm-lit.py: C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\utils\lit\lit\TestingConfig.py:151: fatal: unable to parse config file 'C:/buildbot/as-builder-1/x-armv7l/build/runtimes/runtimes-armv7-unknown-linux-gnueabihf-bins/libunwind/test/cmake-bridge.cfg', traceback: Traceback (most recent call last):
>   File "C:\buildbot\as-builder-1\x-armv7l\llvm-project\llvm\utils\lit\lit\TestingConfig.py", line 139, in load_from_path
>     exec(compile(data, path, "exec"), cfg_globals, None)
>   File "C:/buildbot/as-builder-1/x-armv7l/build/runtimes/runtimes-armv7-unknown-linux-gnueabihf-bins/libunwind/test/cmake-bridge.cfg", line 7
>     config. 'executor="C:/Python310/python.exe" "C:/buildbot/as-builder-1/x-armv7l/llvm-project/libcxx/utils/ssh.py" --host = "ubuntu@jetson6.lab.llvm.org'"
>                                                                                                                                                            ^
> SyntaxError: unterminated string literal (detected at line 7)
> ```
> 
> would you take care of it?
> 
> UPDATE: Just let me know if you will need to check you changes before commit. I will test them on the builder locally.

@vvereschaka Could you try the following patch? It works in my local testing but I don't have the right setup for you toolchain. The problem appears to be that cmake always uses greedy regexes, so I've changed the split logic to instead find the first =.

```
diff --git a/libcxx/test/CMakeLists.txt b/libcxx/test/CMakeLists.txt
index fd194a36a28f..1416a8343df7 100644
--- a/libcxx/test/CMakeLists.txt
+++ b/libcxx/test/CMakeLists.txt
@@ -39,8 +39,10 @@ if (LLVM_USE_SANITIZER)
 endif()
 
 foreach(param IN LISTS LIBCXX_TEST_PARAMS)
-  string(REGEX REPLACE "(.+)=(.+)" "\\1" name "${param}")
-  string(REGEX REPLACE "(.+)=(.+)" "\\2" value "${param}")
+  string(FIND "${param}" "=" _eq_index)
+  string(SUBSTRING "${param}" 0 ${_eq_index} name)
+  string(SUBSTRING "${param}" ${_eq_index} -1 value)
+  string(SUBSTRING "${value}" 1 -1 value) # strip the leading =
   serialize_lit_param("${name}" "\"${value}\"")
 endforeach()
 
diff --git a/libcxxabi/test/CMakeLists.txt b/libcxxabi/test/CMakeLists.txt
index b65ac3a9d164..1baddf9ab305 100644
--- a/libcxxabi/test/CMakeLists.txt
+++ b/libcxxabi/test/CMakeLists.txt
@@ -53,8 +53,10 @@ else()
 endif()
 
 foreach(param IN LISTS LIBCXXABI_TEST_PARAMS)
-  string(REGEX REPLACE "(.+)=(.+)" "\\1" name "${param}")
-  string(REGEX REPLACE "(.+)=(.+)" "\\2" value "${param}")
+  string(FIND "${param}" "=" _eq_index)
+  string(SUBSTRING "${param}" 0 ${_eq_index} name)
+  string(SUBSTRING "${param}" ${_eq_index} -1 value)
+  string(SUBSTRING "${value}" 1 -1 value) # strip the leading =
   serialize_lit_param("${name}" "\"${value}\"")
 endforeach()
 
diff --git a/libunwind/test/CMakeLists.txt b/libunwind/test/CMakeLists.txt
index 084da0ab8f35..9d128720968c 100644
--- a/libunwind/test/CMakeLists.txt
+++ b/libunwind/test/CMakeLists.txt
@@ -36,8 +36,10 @@ else()
 endif()
 
 foreach(param IN LISTS LIBUNWIND_TEST_PARAMS)
-  string(REGEX REPLACE "(.+)=(.+)" "\\1" name "${param}")
-  string(REGEX REPLACE "(.+)=(.+)" "\\2" value "${param}")
+  string(FIND "${param}" "=" _eq_index)
+  string(SUBSTRING "${param}" 0 ${_eq_index} name)
+  string(SUBSTRING "${param}" ${_eq_index} -1 value)
+  string(SUBSTRING "${value}" 1 -1 value) # strip the leading =
   serialize_lit_param("${name}" "\"${value}\"")
 endforeach()

```

vvereschaka commented:
[executorfix.diff.txt](https://github.com/llvm/llvm-project/files/12745676/executorfix.diff.txt)

@arichardson ,

here is few updates for your fix (please see attached file):
* added an escaping of quotes for the serialized lit parameter. There could be an executor string something like that `"C:/Python310/python.exe" "C:/buildbot/temp/llvm-project/libcxx/utils/ssh.py" --host=ubuntu@jetson6.lab.llvm.org`. We need to replace `"` with `\"` because it goes into the python's string.
* the test parameters are processing as a cmake list. We need to properly prepare `xxx_LIBxxx_TEST_PARAMS` variables in the cmake cache file.

The `libunwind` and `libc++abi` tests are passed successfully with these changes. I have started the `libc++` tests. It will take about 1 hour, but they already started successfully. These changes are working properly on the win cross toolchain builders.
